pkg/database/database.go

```diff
@@ -1145,7 +1188,89 @@ func (d *db) serializeTx(tx *store.Tx, spec *schema.EntriesSpec, snap *store.Sna
 	return stx, nil
 }
 
+func (d *db) mayUpdateFollowerState(newState *schema.FollowerState) error {
+	if d.followerStates == nil || newState == nil {
+		return nil
+	}
+
+	// clean up followerStates
+	// it's safe to remove up to latest tx committed in master
+	committedTxID := d.st.LastCommittedTxID()
+
+	for uuid, st := range d.followerStates {
+		if st.precommittedTxID <= committedTxID {
+			delete(d.followerStates, uuid)
+		}
+	}
+
+	// follower pre-committed state should be consistent with master
+	hdr, err := d.st.ReadTxHeader(newState.PrecommittedTxID)
+	if err == store.ErrTxNotFound {
+		return fmt.Errorf("%w: follower is ahead of master", err)
+	}
+	if err != nil {
+		return err
+	}
+
+	newFollowerAlh := schema.DigestFromProto(newState.PrecommittedTxAlh)
+
+	if hdr.Alh() != newFollowerAlh {
+		return fmt.Errorf("%w: follower diverged from master", err)
+	}
+
+	if newState.PrecommittedTxID <= committedTxID {
+		// as far as the master is concerned, nothing really new has happened
+		return nil
+	}
+
+	followerSt, ok := d.followerStates[newState.UUID]
+	if ok {
+		if newState.PrecommittedTxID < followerSt.precommittedTxID {
+			return fmt.Errorf("%w: the newly informed follower state lags behind the previously informed one", ErrIllegalArguments)
+		}
+
+		if newState.PrecommittedTxID == followerSt.precommittedTxID {
+			// as of the last informed follower status update, nothing has changed
+			return nil
+		}
+
+		// actual replication progress is informed by the follower
+		followerSt.precommittedTxID = newState.PrecommittedTxID
+		followerSt.precommittedAlh = newFollowerAlh
+	} else {
+		// follower informs first replication state
+		d.followerStates[newState.UUID] = &followerState{
+			precommittedTxID: newState.PrecommittedTxID,
+			precommittedAlh:  newFollowerAlh,
+		}
+	}
+
+	// check up to which tx enough followers ack replication and it's safe to commit
+	mayCommitUpToTxID := uint64(0)
+	allowances := 0
+
+	// we may clean up followerStates from those who are lag behind commit
+	for _, st := range d.followerStates {
```

Comment:
Just trying to understand this better

allowances will always be equal to the number of followers, this implies acks (allowances>=syncfollowers) from followers will always the same right? Is this to check at least `n replicas` are alive and up to date?